ql/src/java/org/apache/hadoop/hive/ql/metadata/SessionHiveMetaStoreClient.java

```diff
@@ -2079,11 +2091,21 @@ protected GetTableResult getTableInternal(GetTableRequest req) throws TException
     Map<Object, Object> queryCache = getQueryCache();
     if (queryCache != null) {
       // Retrieve or populate cache
+      CacheKey cacheKeyTableId = new CacheKey(KeyType.TABLE_ID,req.getCatName(),req.getDbName(),req.getTblName());
+      long tableId = -1;
+
+      if(queryCache.get(cacheKeyTableId) != null)
+          tableId = (long)queryCache.get(cacheKeyTableId);
+
+      req.setId(tableId);
       CacheKey cacheKey = new CacheKey(KeyType.TABLE, req);
       GetTableResult v = (GetTableResult) queryCache.get(cacheKey);
       if (v == null) {
         v = super.getTableInternal(req);
-        queryCache.put(cacheKey, v);
+        if (tableId == -1)
+          queryCache.put(cacheKeyTableId, v.getTable().getId());
+        else
+          queryCache.put(cacheKey, v);
```

Comment:
Aren't we suppose to add v into queryCache even if tableId == -1? Probably need to set id in req object before updating cache entry.